rcl/test/memory_tools_common.cpp

```diff
@@ -0,0 +1,159 @@
+  }
+  *unexpected_malloc_callback = callback;
+}
+
+void *
+custom_malloc(size_t size)
+{
+  if (!enabled.load()) {return malloc(size);}
+  auto foo = SCOPE_EXIT(enabled.store(true););
+  enabled.store(false);
+  if (!malloc_expected) {
+    if (unexpected_malloc_callback) {
+      (*unexpected_malloc_callback)();
+    }
+  }
+  void * memory = malloc(size);
+  uint64_t fw_size = size;
+  MALLOC_PRINTF(
+    "malloc  expected(%s): %p %" PRIu64 "\n",
+    malloc_expected ? "true " : "false", memory, fw_size);
+  return memory;
+}
+
+static __thread bool realloc_expected = true;
+static __thread UnexpectedCallbackType * unexpected_realloc_callback = nullptr;
+void set_on_unepexcted_realloc_callback(UnexpectedCallbackType callback)
+{
+  if (unexpected_realloc_callback) {
+    unexpected_realloc_callback->~UnexpectedCallbackType();
+    free(unexpected_realloc_callback);
+    unexpected_realloc_callback = nullptr;
+  }
+  if (!callback) {
+    return;
+  }
+  if (!unexpected_realloc_callback) {
+    unexpected_realloc_callback =
+      reinterpret_cast<UnexpectedCallbackType *>(malloc(sizeof(UnexpectedCallbackType)));
+    if (!unexpected_realloc_callback) {
+      throw std::bad_alloc();
+    }
+    new (unexpected_realloc_callback) UnexpectedCallbackType();
+  }
+  *unexpected_realloc_callback = callback;
+}
+
+void *
+custom_realloc(void * memory_in, size_t size)
+{
+  if (!enabled.load()) {return realloc(memory_in, size);}
+  auto foo = SCOPE_EXIT(enabled.store(true););
+  enabled.store(false);
+  if (!realloc_expected) {
+    if (unexpected_realloc_callback) {
+      (*unexpected_realloc_callback)();
+    }
+  }
+  void * memory = realloc(memory_in, size);
+  uint64_t fw_size = size;
+  MALLOC_PRINTF(
+    "realloc expected(%s): %p %p %" PRIu64 "\n",
+    malloc_expected ? "true " : "false", memory_in, memory, fw_size);
+  return memory;
+}
+
+static __thread bool free_expected = true;
+static __thread UnexpectedCallbackType * unexpected_free_callback = nullptr;
+void set_on_unepexcted_free_callback(UnexpectedCallbackType callback)
+{
+  if (unexpected_free_callback) {
+    unexpected_free_callback->~UnexpectedCallbackType();
+    free(unexpected_free_callback);
+    unexpected_free_callback = nullptr;
+  }
+  if (!callback) {
+    return;
+  }
+  if (!unexpected_free_callback) {
+    unexpected_free_callback =
+      reinterpret_cast<UnexpectedCallbackType *>(malloc(sizeof(UnexpectedCallbackType)));
+    if (!unexpected_free_callback) {
+      throw std::bad_alloc();
+    }
+    new (unexpected_free_callback) UnexpectedCallbackType();
+  }
+  *unexpected_free_callback = callback;
+}
+
+void
+custom_free(void * memory)
+{
+  if (!enabled.load()) {return free(memory);}
+  auto foo = SCOPE_EXIT(enabled.store(true););
+  enabled.store(false);
+  if (!free_expected) {
+    if (unexpected_free_callback) {
+      (*unexpected_free_callback)();
+    }
+  }
+  MALLOC_PRINTF("free    expected(%s): %p\n", malloc_expected ? "true " : "false", memory);
```

Comment:
the spaces after `free` look a little weird. Maybe do something like `free():    expected...`